### PR TITLE
fixes #23438 - pre/post cv promote hooks

### PR DESCRIPTION
**Pull request by thomasmckay**

### CODE REVIEW

**Comment by thomasmckay** [comment]
@jlsherrill - piggy-backed on your PR (unnecessarily but it was the inspiration)

**Comment by jturel** [comment]
@thomasmckay - some test failures here

**Comment by thomasmckay** [comment]
@jlsherrill - I thought this was working but it's not. Any pointers?
```
$ rake hooks:objects
/home/vagrant/code/foreman/lib/foreman.rb:8: warning: already initialized constant Foreman::UUID_REGEXP
/home/vagrant/code/foreman/lib/foreman.rb:8: warning: previous definition of UUID_REGEXP was here
2018-05-02T12:30:05 [D|app|] Found hook to Katello::Repository#after_sync, filename signatures_hook.rb
2018-05-02T12:30:05 [D|app|] Found hook to Katello::ContentViewVersion#after_promote, filename signatures_hook.rb
2018-05-02T12:30:05 [D|app|] Found hook to Katello::ContentViewVersion#before_promote, filename signatures_hook.rb
2018-05-02T12:30:05 [I|app|] Finished discovering 1 hooks for Katello::Repository#after_sync
2018-05-02T12:30:05 [I|app|] Finished discovering 1 hooks for Katello::ContentViewVersion#after_promote
2018-05-02T12:30:05 [I|app|] Finished discovering 1 hooks for Katello::ContentViewVersion#before_promote
2018-05-02T12:30:23 [D|app|] Extending Katello::ContentViewVersion with foreman_hooks Rails hooking support
2018-05-02T12:30:23 [D|app|] Created hook method after_promote on Katello::ContentViewVersion
2018-05-02T12:30:23 [E|app|] foreman_hooks: unknown hook object Katello::ContentViewVersion, check against `foreman-rake hooks:objects`
```
Here's my config dir
```
$ tree config/hooks/
config/hooks/
└── katello
    ├── content_view_version
    │   ├── after_promote
    │   │   └── signatures_hook.rb
    │   └── before_promote
    │       └── signatures_hook.rb
    └── repository
        └── after_sync
            └── signatures_hook.rb
```

**Comment by jlsherrill** [comment]
you're missing something like: https://github.com/Katello/katello/blob/master/app/models/katello/repository.rb#L39

**Comment by thomasmckay** [comment]
@jlsherrill This all seems to work in that I hit breakpoints in my hook methods and see the hooks loaded at server startup... but the script placed is never executed. It works fine for repo after_sync.

File: app/models/katello/content_view_version.rb [hunk 1]
```diff
@@ -348,6 +350,24 @@ def validate_destroyable!(skip_environment_check = false)
...
+        true
+      end
+    end
+
+    def after_promote_hook
+      run_callbacks :after_promote do
+        logger.debug "custom hook after_promote on #{name} will be executed if defined."
+        true
+      end
+    end
```
Comment (jlsherrill) [inline app/models/katello/content_view_version.rb]: replace both of these methods with:
```
    def promote_hook
      run_callbacks :promote do
        logger.debug "custom hook promote on #{name} will be executed if defined."
        true
      end
    end
```
and then you call:
`  ::Katello::ContentViewVersion.find(input[:id]).after_promote_hooks `  NOTE the plural hooks

**Comment by thomasmckay** [comment]
@jlsherrill victory! thanks

File: app/services/katello/puppet_class_importer_extensions.rb [hunk 2]
```diff
@@ -3,32 +3,30 @@ module Services
...
+          changed[kind].slice!(@environment) unless changed[kind].empty?
+        end
 
-          #prevent the puppet environment from being deleted, by removing special '_destroy_' String
-          if changed['obsolete'][@environment]
-            changed['obsolete'][@environment] =
+        #prevent the puppet environment from being deleted, by removing special '_destroy_' String
+        if changed['obsolete'][@environment]
+          changed['obsolete'][@environment] =
               changed['obsolete'][@environment].select { |klass| klass != '_destroy_' }
```
Comment (jlsherrill) [inline app/services/katello/puppet_class_importer_extensions.rb]: Rubocop failure: 

/var/lib/workspace/workspace/katello-pr-test/app/services/katello/puppet_class_importer_extensions.rb:17:15: C: Layout/IndentAssignment: Indent the first line of the right-hand-side of a multi-line assignment.

              changed['obsolete'][@environment].select { |klass| klass != '_destroy_' }

              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Comment by ekohl** [comment]
This caused a regression in our release pipeline. I've created https://projects.theforeman.org/issues/23636 but for completeness also the error here:
```
not ok 13 promote content view
# (in test file fb-content-katello.bats, line 220)
#   `hammer content-view version promote  --organization="${ORGANIZATION}" \' failed with status 70
# Task 5778f221-9605-495a-a8bb-d82e07f91045 running: 0.0/1, 0%, elapsed: 00:00:00
# Task 5778f221-9605-495a-a8bb-d82e07f91045 running: 0.46153846153846156/1, 46%, 0.2/s, elapsed: 00:00:02, ETA: 00:00:02
# Task 5778f221-9605-495a-a8bb-d82e07f91045 error: 0.6538461538461539/1, 65%, 0.2/s, elapsed: 00:00:04, ETA: 00:00:02
# Task 5778f221-9605-495a-a8bb-d82e07f91045 error: 0.6538461538461539/1, 65%, 0.2/s, elapsed: 00:00:04, ETA: 00:00:02
# Error: undefined method `before_promote_hooks' for #<Katello::ContentViewVersion:0x00000000090a98e0>
# Did you mean?  before_remove_for_history
```

**Pull request merged**